newm-server/src/main/kotlin/io/newm/server/features/marketplace/repo/MarketplaceRepositoryImpl.kt

```diff
@@ -141,6 +130,114 @@ internal class MarketplaceRepositoryImpl(
+        log.debug { "generateCreateSaleAmount: $request" }
+
+        val nativeAssets =
+            listOf(
+                nativeAsset {
+                    policy = request.bundlePolicyId
+                    name = request.bundleAssetName
+                    amount = computeAmount(request.bundleAmount, request.totalBundleQuantity)
+                }
+            )
+
+        val lovelace = configRepository.getLong(CONFIG_KEY_MARKETPLACE_SALE_LOVELACE)
+        val amount =
+            CborArray.create(
+                listOf(
+                    CborInteger.create(lovelace.toBigInteger()),
+                    nativeAssets.toNativeAssetCborMap()
+                )
+            )
+
+        val pendingSaleTimeToLive = configRepository.getLong(CONFIG_KEY_MARKETPLACE_PENDING_SALE_TTL)
+        val costPolicyId = request.costPolicyId ?: configRepository.getString(CONFIG_KEY_MARKETPLACE_CURRENCY_POLICY_ID)
+        val costAssetName =
+            request.costAssetName ?: configRepository.getString(CONFIG_KEY_MARKETPLACE_CURRENCY_ASSET_NAME)
+        val sale =
+            transaction {
+                MarketplacePendingSaleEntity.deleteAllExpired(pendingSaleTimeToLive)
+                MarketplacePendingSaleEntity.new {
+                    this.ownerAddress = request.ownerAddress
+                    this.bundlePolicyId = request.bundlePolicyId
+                    this.bundleAssetName = request.bundleAssetName
+                    this.bundleAmount = request.bundleAmount
+                    this.costPolicyId = costPolicyId
+                    this.costAssetName = costAssetName
+                    this.costAmount = request.costAmount
+                    this.totalBundleQuantity = request.totalBundleQuantity
+                }
+            }
+
+        return SaleAmountResponse(
+            saleId = sale.id.value,
+            amountCborHex = amount.toCborByteArray().toHexString()
+        )
+    }
+
+    override suspend fun generateSaleTransaction(request: SaleTransactionRequest): SaleTransactionResponse {
+        log.debug { "generateSaleTransaction: $request" }
+
+        if (request.utxos.isEmpty()) {
+            throw HttpPaymentRequiredException("Missing UTXO's")
+        }
+
+        val pendingSaleTimeToLive = configRepository.getLong(CONFIG_KEY_MARKETPLACE_PENDING_SALE_TTL)
+        val sale =
+            transaction {
+                MarketplacePendingSaleEntity.deleteAllExpired(pendingSaleTimeToLive)
+                MarketplacePendingSaleEntity[request.saleId]
+            }
+
+        val contractAddress = configRepository.getString(CONFIG_KEY_MARKETPLACE_SALE_CONTRACT_ADDRESS)
+        val lovelace = configRepository.getLong(CONFIG_KEY_MARKETPLACE_SALE_LOVELACE)
+        val transaction =
+            cardanoRepository.buildTransaction {
+                sourceUtxos.addAll(request.utxos)
+                outputUtxos.add(
+                    outputUtxo {
```

Comment:
If we are starting the sale immediately, we also need to mint the starter token with this transaction. This makes the transaction more complicated because it needs to execute the smart contract to do that and requires the NEWM hot key to sign it.

If we are always scheduling the start, we need to use the scheduler system and create a new Job to start the sale. We can ensure we ask for enough funds in this TX to make sure we can pay for the next.

All this is for the next PR, but I wanted to mention it here.